### ISSUE TITLE
Molmo2 MLX Implementation Produces Broken Output (Repeats Prompts)

### ISSUE THREAD

**Issue by avrahamraskin** [Blaizzy/mlx-vlm#655]
## Summary

The Molmo2 implementation in mlx-vlm produces broken output - models just repeat the input prompt instead of generating meaningful responses. This affects **all** Molmo2 MLX models tested (community conversions and custom conversions using mlx-vlm's convert tool).

## Environment

- **mlx-vlm version**: 0.3.10 (commit fc8c92e)
- **Hardware**: M2 Mac, 16GB RAM
- **macOS**: 26.2 (25C5089a)
- **Python**: 3.13
- **MLX**: 0.29.4

### Test 1: Community Model (Aliyovic/molmo2-4b-mlx-8bit)

```python
from mlx_vlm import load, generate

model, processor = load("Aliyovic/molmo2-4b-mlx-8bit")
output = generate(model, processor, "What is 2+2?", max_tokens=50, verbose=True)
print(output)
```

**Expected:** Meaningful answer (e.g., "4" or "The answer is 4")
**Actual:** `What is 2+2?What is  2+2?What is  2+2?What is  2+2?...` (repeats prompt)

### Test 2: Custom FP16 Conversion

```bash
poetry run python -m mlx_vlm convert \
  --hf-path allenai/Molmo2-4B \
  --mlx-path ./molmo2-4b-mlx-fp16 \
  --dtype float16
```

Then load and test:

```python
model, processor = load("./molmo2-4b-mlx-fp16")
output = generate(model, processor, "What is 2+2?", max_tokens=50)
print(output)
```

**Result:** Identical broken behavior - repeats prompt

### Test 3: Custom 4-bit Quantized Conversion

```bash
poetry run python -m mlx_vlm convert \
  --hf-path allenai/Molmo2-4B \
  --mlx-path ./molmo2-4b-mlx-4bit \
  --quantize --q-bits 4 --q-group-size 64
```

**Result:** 
1. Conversion succeeds (model quantized to 9.614 bits per weight)
2. Cannot load with mlx-vlm's `load()` - architecture mismatch (290 extra quantization parameters)
3. When loading weights manually, same prompt repetition issue

## Evidence All Three Produce Same Broken Behavior

| Model | Source | Size | Loads | Output Behavior |
|-------|--------|------|-------|-----------------|
| Aliyovic/molmo2-4b-mlx-8bit | Community | ~5GB | ✅ | ❌ Repeats prompt |
| Custom FP16 | mlx-vlm convert | ~9GB | ✅ | ❌ Repeats prompt |
| Custom 4-bit | mlx-vlm convert | ~5.4GB | ⚠️ Arch mismatch | ❌ (Same issue) |

**Identical output from all three:**
```
Prompt: What is 2+2?
Output: What is 2+2?What is  2+2?What is  2+2?What is  2+2?What is   2+2?...
```

## Analysis

Since **all three independent conversions** produce identical broken behavior, the issue is not with:
- Weight conversion process ✅
- Quantization ✅
- Model size/memory ✅
- Community conversion quality ✅

**The bug is in mlx-vlm's Molmo2 implementation itself**, likely in:
- `mlx_vlm/models/molmo2/molmo2.py` - forward pass or generation logic
- `mlx_vlm/models/molmo2/language.py` - text generation
- Special token handling (frame tokens, image tokens)
- Attention mask computation
- KV cache management

## Additional Observations

1. **Model loads successfully** - no errors during loading
2. **Memory usage is reasonable** - ~10.6GB peak for FP16 (fits in 16GB)
3. **Generation speed is normal** - 8-10 tokens/sec
4. **Only output is broken** - everything else works

## PyTorch Baseline

The official PyTorch version (`allenai/Molmo2-4B`) works correctly but:
- Molmo2-8B: OOM on 16GB M2 (exceeds memory)
- Molmo2-4B: Hangs silently after model load (MPS backend incompatibility)

This confirms PyTorch MPS doesn't work with Molmo2, making MLX the only viable path for Apple Silicon.

## Related Issues/PRs

- Issue #39: Lists "Molmo" as model to port
- v0.3.2 release notes: "Fix for phi3_v and molmo mask"

Molmo2 appears to be a recent addition (different from original Molmo), and may not have been fully tested.

## Request

Could the maintainers:
1. Confirm if Molmo2 support is complete/tested?
2. Investigate the prompt repetition issue?
3. Provide guidance on proper Molmo2 usage with mlx-vlm?

I'm happy to help test fixes or provide additional debugging information.

## Test Scripts

Complete test scripts and outputs available at:
https://github.com/avrahamraskin/orin-security-system/tree/main

Files:
- `MOLMO2_MLX_TEST_RESULTS.md` - Community model testing
- `MOLMO2_MLX_CONVERSION_RESULTS.md` - Custom conversion results
- `test_molmo2_simple.py` - Minimal reproduction
- `test_molmo2_mlx_community.py` - Full test suite

## Why This Matters

Molmo2 is state-of-the-art for video understanding (62.8 benchmark) and would be valuable for Apple Silicon users, but the current implementation prevents its use. Getting this fixed would make Molmo2 the best video VLM option for MLX.

Thank you for maintaining mlx-vlm! 🙏
